build_tools/cmake/iree_alwayslink.cmake

```diff
@@ -12,30 +12,80 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+include(CMakeParseArguments)
+
 # Additional libraries containing statically registered functions/flags, which
 # should always be linked in to binaries.
 
-set(_ALWAYSLINK_LIBS
-  MLIRAffineOps
-  MLIRAnalysis
-  MLIREDSC
-  MLIRIR
-  MLIRParser
-  MLIRPass
-  MLIRSPIRV
-  MLIRSPIRVSerialization
-  MLIRSPIRVTransforms
-  MLIRStandardOps
-  MLIRTransforms
-  MLIRTranslation
-  MLIRSupport
-  MLIRVectorOps
-  MLIRLinalgOps
-  tensorflow::mlir_xla
-)
-
-foreach(LIB ${_ALWAYSLINK_LIBS})
-  if(TARGET LIB)
-    set_property(TARGET ${LIB} PROPERTY ALWAYSLINK 1)
-  endif()
-endforeach(LIB)
+
+# set_alwayslink_property()
+#
+# CMake function to set the ALWAYSLINK on external libraries
+#
+# Parameters:
+# ALWAYSLINK_LIBS: List of libraries
+# SKIP_NONEXISTING: When added, ALWAYSLINK is only set on existing libraries.
+
+function(set_alwayslink_property)
+  cmake_parse_arguments(
+    _RULE
+    "SKIP_NONEXISTING"
+    ""
+    "ALWAYSLINK_LIBS"
+    ${ARGN}
+  )
+
+  foreach(_LIB ${_RULE_ALWAYSLINK_LIBS})
+    # If SKIP_NONEXISTING is false: Always try to set the property.
+    # If SKIP_NONEXISTING is true : Only set the property if the target exists.
+    if((_RULE_SKIP_NONEXISTING AND TARGET ${_LIB}) OR NOT _RULE_SKIP_NONEXISTING)
```

Comment:
I found this logic a bit difficult to follow. How about this instead?

```
foreach(_LIB ${_RULE_ALWAYSLINK_LIBS})
  if(NOT TARGET ${_LIB} AND _RULE_SKIP_NONEXISTING)
    continue()
  endif()

  # Check if the target is an aliased target.
  ...
endforeach()
```

Reply:
Nice! I wasn't aware of `continue()`. Changed it.